File: homework/ba-nguyen/lesson-7/longest-zigzag-path-in-a-binary-tree.ts
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function longestZigZag(root: TreeNode | null): number {
  let res = 0;

  function dfs(root: TreeNode | null) {
    if (!root) return [-1, -1];

    const left = dfs(root.left);
    const right = dfs(root.right);
    const pathLeft = left[1] + 1;
    const pathRight = right[0] + 1;

    res = Math.max(res, pathLeft || res, pathRight);

    return [pathLeft, pathRight];
  }
  dfs(root);

  return res;
}

function longestZigZag1(root: TreeNode | null): number {
  let res = 0;

  function dfs(root: TreeNode | null, direction: string, step: number) {
    if (!root) return;

    res = Math.max(res, step);

    if (direction === "left") {
      dfs(root.left, "left", 1);
      dfs(root.right, "right", step + 1);
    } else {
      dfs(root.right, "right", 1);
      dfs(root.left, "left", step + 1);
    }
  }

  dfs(root, "left", 0);
  dfs(root, "right", 0);

  return res;
}

File: homework/quang_tu/lesson_5/longest-zigzag-path-in-a-binary-tree.cpp
class Solution {
    int maxLength = 0;
public:
    int longestZigZag(TreeNode* root) {
        findZigZag(root->left, false, 1);
        findZigZag(root->right, true, 1);
        return maxLength;
    }

    void findZigZag(TreeNode* root, bool side, int sum) {
        if (!root)
            return;

        maxLength = max(maxLength, sum);
        if (side) {
            if (root->left)
                findZigZag(root->left, false, sum+1);

            findZigZag(root->right, true, 1);
        } else {
            if (root->right)
                findZigZag(root->right, true, sum+1);

            findZigZag(root->left, false, 1);
        }
    }
};
